docs/ext/pydantic_autosummary/__init__.py

```diff
@@ -0,0 +1,921 @@
+
+
+class AutoLink(SphinxRole):
+    """Smart linking role.
+
+    Expands to ':obj:`text`' if `text` is an object that can be imported;
+    otherwise expands to '*text*'.
+    """
+
+    def run(self) -> tuple[list[Node], list[system_message]]:
+        pyobj_role = self.env.get_domain("py").role("obj")
+        assert pyobj_role is not None
+        objects, errors = pyobj_role(
+            "obj",
+            self.rawtext,
+            self.text,
+            self.lineno,
+            self.inliner,
+            self.options,
+            self.content,
+        )
+        if errors:
+            return objects, errors
+
+        assert len(objects) == 1
+        pending_xref = cast(addnodes.pending_xref, objects[0])
+        try:
+            # try to import object by name
+            prefixes = get_import_prefixes_from_env(self.env)
+            import_by_name(pending_xref["reftarget"], prefixes)
+        except ImportExceptionGroup:
+            literal = cast(nodes.literal, pending_xref[0])
+            objects[0] = nodes.emphasis(
+                self.rawtext, literal.astext(), classes=literal["classes"]
+            )
+
+        return objects, errors
+
+
+def get_rst_suffix(app: Sphinx) -> str | None:
+    def get_supported_format(suffix: str) -> tuple[str, ...]:
+        parser_class = app.registry.get_source_parsers().get(suffix.removeprefix("."))
+        if parser_class is None:
+            return ("restructuredtext",)
+        return parser_class.supported
+
+    suffix = None
+    for suffix in app.config.source_suffix:
+        if "restructuredtext" in get_supported_format(suffix):
+            return suffix
+
+    return None
+
+
+def process_generate_options(app: Sphinx) -> None:
+    genfiles = app.config.autosummary_generate
+
+    if genfiles is True:
+        env = app.builder.env
+        genfiles = [
+            env.doc2path(x, base=False)
+            for x in env.found_docs
+            if os.path.isfile(env.doc2path(x))
+        ]
+    elif genfiles is False:
+        pass
+    else:
+        ext = list(app.config.source_suffix)
+        genfiles = [
+            genfile + (ext[0] if not genfile.endswith(tuple(ext)) else "")
+            for genfile in genfiles
+        ]
+
+        for entry in genfiles[:]:
+            if not path.isfile(path.join(app.srcdir, entry)):
+                logger.warning(__("autosummary_generate: file not found: %s"), entry)
+                genfiles.remove(entry)
+
+    if not genfiles:
+        return
+
+    suffix = get_rst_suffix(app)
+    if suffix is None:
+        logger.warning(
+            __(
+                "autosummary generates .rst files internally. "
+                "But your source_suffix does not contain .rst. Skipped."
+            )
+        )
+        return
+
+    # ----------- pydantic_autosummary change
+    from .generate import generate_autosummary_docs
+    # ----------/ pydantic_autosummary change
```

Comment:
This file is a direct copy of autosummary, and this is the only change in it (it was previously `sphinx.autosummary.generate` or so), therefore I don't think the rest of this file needs to be reviewed.

There were very minor changes just to adapt to our ruff linting.